### PR TITLE
NoAliasFunctionsFixer - Add more function aliases

### PR DESCRIPTION
**Pull request by danog**
Extracted directly from the PHP trunk (master&7.4).

### CODE REVIEW

**Comment by danog** [comment]
Note quite sure what to make of the phpstan analysis, seems like a bug to me (mixed is a supertype of string; plus the error is not triggered on other lines of the elseif))
As a friendly recommendation from a psalm lover, I can really recommend switching to psalm.dev for static analysis :)

**Comment by kubawerlos** [comment]
That's weird, I've reported a bug: https://github.com/phpstan/phpstan/issues/4793 - check out the example, it's really strange.

As workaround you can add above the line with `foreach ($this->configuration['sets'] as $set) {` PHPdoc like this: /** @var string $set */

> I can really recommend switching to https://psalm.dev

Why not both? I often use both as the can detect different issues and both have some bugs.

**Comment by GrahamCampbell** [comment]
Needs tests covering some of these options.

File: src/Fixer/Alias/NoAliasFunctionsFixer.php [hunk 1]
```diff
@@ -245,13 +351,36 @@ protected function applyFix(\SplFileInfo $file, Tokens $tokens)
...
+            '@openssl' => 'openssl functions',
+            '@sodium' => 'libsodium functions',
+            '@ftp' => 'FTP functions',
+            '@posix' => 'POSIX functions',
+            '@pcntl' => 'PCNTL functions',
+        ];
+
+        $list = '';
+        foreach ($sets as $set => $description) {
+            $list .= "* `{$set}` ({$description})\n";
```
Comment (SpacePossum) [inline src/Fixer/Alias/NoAliasFunctionsFixer.php:376]: please use `sprintf`

File: src/Fixer/Alias/NoAliasFunctionsFixer.php [hunk 2]
```diff
@@ -245,13 +351,36 @@ protected function applyFix(\SplFileInfo $file, Tokens $tokens)
...
+        }
 
         return new FixerConfigurationResolver([
-            (new FixerOptionBuilder('sets', 'List of sets to fix. Defined sets are `@internal` (native functions), `@IMAP` (IMAP functions), `@mbreg` (from `ext-mbstring`) `@all` (all listed sets).'))
+            (new FixerOptionBuilder('sets', "List of sets to fix. Defined sets are:\n\n{$list}"))
```
Comment (SpacePossum) [inline src/Fixer/Alias/NoAliasFunctionsFixer.php:380]: please use `sprintf`

File: src/Fixer/Alias/NoAliasFunctionsFixer.php [hunk 3]
```diff
@@ -245,13 +351,36 @@ protected function applyFix(\SplFileInfo $file, Tokens $tokens)
      */
     protected function createConfigurationDefinition()
     {
-        $sets = ['@internal', '@IMAP', '@mbreg', '@all', '@time', '@exif'];
+        $sets = [
+            '@all' => 'all listed sets',
```
Comment (SpacePossum) [inline src/Fixer/Alias/NoAliasFunctionsFixer.php:355]: please sort the list in alphabetical order

File: src/Fixer/Alias/NoAliasFunctionsFixer.php [hunk 4]
```diff
@@ -245,13 +351,36 @@ protected function applyFix(\SplFileInfo $file, Tokens $tokens)
...
+        }
 
         return new FixerConfigurationResolver([
-            (new FixerOptionBuilder('sets', 'List of sets to fix. Defined sets are `@internal` (native functions), `@IMAP` (IMAP functions), `@mbreg` (from `ext-mbstring`) `@all` (all listed sets).'))
+            (new FixerOptionBuilder('sets', "List of sets to fix. Defined sets are:\n\n{$list}"))
                 ->setAllowedTypes(['array'])
-                ->setAllowedValues([new AllowedValueSubset($sets)])
-                ->setDefault(['@internal', '@IMAP'])
+                ->setAllowedValues([new AllowedValueSubset(array_keys($sets))])
+                ->setDefault(['@internal', '@IMAP', '@pg'])
```
Comment (SpacePossum) [inline src/Fixer/Alias/NoAliasFunctionsFixer.php:383]: please don't change the default

File: src/Fixer/Alias/NoAliasFunctionsFixer.php [hunk 5]
```diff
@@ -107,6 +180,17 @@ public function configure(array $configuration = null)
                 $this->aliases = array_merge($this->aliases, self::$mbregSet);
                 $this->aliases = array_merge($this->aliases, self::$timeSet);
                 $this->aliases = array_merge($this->aliases, self::$exifSet);
+                $this->aliases = array_merge($this->aliases, self::$snmpSet);
```
Comment (SpacePossum) [inline src/Fixer/Alias/NoAliasFunctionsFixer.php:183]: please sort in alphabetical order, also the if statements from line 198 till 229

**Comment by SpacePossum** [comment]
Thanks @danog for these improvements :)

Please add the some tests [here](https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/v2.19.0/tests/Fixer/Alias/NoAliasFunctionsFixerTest.php) and have a look at my minors, looking close to finished

**Comment by danog** [comment]
Finally got around to finishing this MR :)
Applied the requested changes, and added some tests based on the existing test logic.
The only detail that I'm not sure about is the `@pg` set: the only alias function contained in that set is the extremely old `pg_exec` alias, deprecated way back in PHP 4.
I'm not sure if it's worth removing from the defaults.

File: src/Fixer/Alias/NoAliasFunctionsFixer.php (hunk 4 above)
Comment (danog) [reply src/Fixer/Alias/NoAliasFunctionsFixer.php:384]: I'm not quite sure about this set: the only alias function contained in the `@pg` is the extremely old pg_exec alias, deprecated way back in PHP 4.
I'm not sure if it's worth removing from the defaults.

File: src/Fixer/Alias/NoAliasFunctionsFixer.php [hunk 6]
```diff
@@ -103,24 +176,57 @@ public function configure(array $configuration = null)
...
+                $this->aliases = array_merge($this->aliases, self::$sodiumSet);
                 $this->aliases = array_merge($this->aliases, self::$timeSet);
-                $this->aliases = array_merge($this->aliases, self::$exifSet);
 
                 break;
             }
 
             if ('@internal' === $set) {
                 $this->aliases = array_merge($this->aliases, self::$internalSet);
+            } elseif ('@exif' === $set) {
```
Comment (kubawerlos) [inline src/Fixer/Alias/NoAliasFunctionsFixer.php:200]: Even before this PR the `if`/`elseif` block wasn't looking pretty - can we have an array `@setname` => self::$setValues` and iterate over it?

Comment (danog) [reply src/Fixer/Alias/NoAliasFunctionsFixer.php]: Done!

**Comment by danog** [comment]
All OK over here? :)

**Review by SpacePossum — APPROVED** [review]

File: src/Fixer/Alias/NoAliasFunctionsFixer.php [hunk 7]
```diff
@@ -30,98 +30,150 @@
  */
 final class NoAliasFunctionsFixer extends AbstractFixer implements ConfigurationDefinitionFixerInterface
 {
-    /** @var array<string, array<int|string>|string> stores alias (key) - master (value) functions mapping */
-    private $aliases = [];
+    private const SETS = [
```
Comment (keradus) [inline src/Fixer/Alias/NoAliasFunctionsFixer.php:33]:
```suggestion
    private const SETS = [ 
```

**Comment by keradus** [comment]
@danog can you rebase on 2.19 or allow to modify your branch by us? we need to ensure that CI kicks off for this PR and all jobs to be green, before we are able to merge.

**Comment by keradus** [comment]
I don't see it solving any bug in the existing fixer. it's giving existing fixer more functionality (to handle more aliases)

should go to 3.1 line

**Comment by danog** [comment]
There, all fixed up nicely.

**Comment by keradus** [comment]
Thank you @danog.

**Pull request closed without merging**